plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -558,25 +558,48 @@ lookupVarName varName@(NamedDeBruijn _ varIx) varEnv =
             var = Var () varName
         Just val -> pure val
 
+{- Note [Laziness of evalBuiltinApp]
+The last three arguments constitute a 'BuiltinRuntime', however
+
+1. if the 'RuntimeScheme' is a 'RuntimeSchemeResult', then we got our result and don't need to
+   construct a 'VBuiltin' value and so don't need to construct a 'BuiltinRuntime' for it, hence
+   constructing it prematurely would be wasteful
+2. more importantly, a 'BuiltinRuntime' is strict in all of its arguments, so constructing one
+   before this function is called would force the result of the builtin application before costing
+   is done for it, while would be very wrong: we don't want to let the user run something that they
+   potentially don't have the budget for
+
+hence the arguments are passed separately instead of being assembled in a 'BuiltinRuntime'.
+
+If you refactor 'evalBuiltinApp', ensure that costing is done before the result is forced. If you
+break this, you'll get a test failure.
+-}
+
+-- See Note [Laziness of evalBuiltinApp].
 -- | Take pieces of a possibly partial builtin application and either create a 'CekValue' using
 -- 'makeKnown' or a partial builtin application depending on whether the built-in function is
 -- fully saturated or not.
 evalBuiltinApp
     :: (GivenCekReqs uni fun s, PrettyUni uni fun)
     => fun
     -> Term NamedDeBruijn uni fun ()
-    -> BuiltinRuntime (CekValue uni fun)
+    -> RuntimeScheme n
+    -> ToRuntimeDenotationType (CekValue uni fun) n
+    -> ToCostingType n
     -> CekM uni fun s (CekValue uni fun)
-evalBuiltinApp fun term runtime@(BuiltinRuntime sch getX cost) = case sch of
+evalBuiltinApp fun term sch getX cost = case sch of
     RuntimeSchemeResult -> do
+        -- See Note [Laziness of evalBuiltinApp].
         spendBudgetCek (BBuiltinApp fun) cost
         case getX of
             MakeKnownFailure logs err       -> do
                 ?cekEmitter logs
                 throwKnownTypeErrorWithCause term err
             MakeKnownSuccess x              -> pure x
             MakeKnownSuccessWithLogs logs x -> ?cekEmitter logs $> x
-    _ -> pure $ VBuiltin fun term runtime
+    -- We _are_ going to need this value, hence we construct it strictly. Otherwise its construction
+    -- is pointlessly delayed, which is clearly visible in Core.
+    _ -> pure $! VBuiltin fun term (BuiltinRuntime sch getX cost)
```

Comment:
This mostly matches with the previous version where `evalBuiltinApp` expected a `BuiltinRuntime`. Previously we would only construct the `BuiltinRuntime` eagerly and now it's the entire `VBuiltin`.

Reply:
Does the `$!` make a difference here? The argument is already in WHNF.

Reply:
It's not! It's only in WHNF as long as each of its components is in WHNF.

Here's what we get without the bang:

```haskell
jump $wreturnCek_sHkO
  ww_sHkM
  ctx_an34
  (case y_an3C of dt14_XqLa { __DEFAULT ->
   VBuiltin
     fun2_an3s
     <...>
     (BuiltinRuntime
        wild16_XbZ dt14_XqLa vx_iu5D)
   })
  w2_sHkJ;
```

and with the bang:

```haskell
case y_an3C of dt14_XqLe { __DEFAULT ->
  jump $wreturnCek_sHl4
    ww_sHl2
    ctx_an34
    (VBuiltin
       fun2_an3s
       <...>
       (BuiltinRuntime
          wild16_XbZ dt14_XqLe vx_itoO))
    w2_sHkZ
```

The former pointlessly defers forcing of `getX`.

Benchmarking results are inconclusive as to whether this costs us something or not (with the bang the code is 0.6-1% faster, but it's below our noise threshold). But I'd rather stay on the safe side.

Reply:
Oh yeah you are right, I looked at the wrong `VBuiltIn` (the one in Ck) whose fields are lazy.